bpf/include/bpf/ctx/skb.h

```diff
@@ -56,6 +57,37 @@
 #define get_hash(ctx)		ctx->hash
 #define get_hash_recalc(ctx)	get_hash(ctx)
 
+#if !defined(__non_bpf_context) && defined(__bpf__)
+#define __CTX_OFF_MAX			0xff
+static __always_inline __maybe_unused int
+skb_load_bytes(const struct __sk_buff *skb, __u64 off, void *to, const __u64 len)
+{
+	void *from;
+	int ret;
+	/* LLVM tends to generate code that verifier doesn't understand,
+	 * so force it the way we want it in order to open up a range
+	 * on the reg.
+	 */
+	asm volatile("r1 = *(u32 *)(%[skb] +76)\n\t"
+		     "r2 = *(u32 *)(%[skb] +80)\n\t"
+		     "%[off] &= %[offmax]\n\t"
+		     "r1 += %[off]\n\t"
+		     "%[from] = r1\n\t"
+		     "r1 += %[len]\n\t"
+		     "if r1 > r2 goto +2\n\t"
+		     "%[ret] = 0\n\t"
+		     "goto +1\n\t"
+		     "%[ret] = %[errno]\n\t"
+		     : [ret]"=r"(ret), [from]"=r"(from)
+		     : [skb]"r"(skb), [off]"r"(off), [len]"ri"(len),
+		       [offmax]"i"(__CTX_OFF_MAX), [errno]"i"(-EINVAL)
+		     : "r1", "r2");
+	if (!ret)
+		memcpy(to, from, len);
+	return ret;
+}
+#endif
```

Comment:
Unfortunately this is not a direct equivalent of `bpf_skb_load_bytes()`. The reason is that the latter uses `skb_header_pointer()` which can query the skb linear data or non-linear parts. The latter is done via `skb_copy_bits ()` without pulling data into linear section.

 - https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/core/filter.c#n1720
 - https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/skbuff.h#n3681

The implementation above is only looking at the linear data, but not non-linear one. In case of XDP, there is only linear data (but not non-linear one, hence the asm inline equivalent in case of XDP).

The case of https://github.com/cilium/cilium/issues/10808 is saying that instead of implementing `skb_load_bytes()` via inline asm where internally does the `memcpy()` in your case, we should rather successively replace individual calls to `skb_load_bytes()`(here: meaning helper call) with direct `ctx->data` access like elsewhere in our code.

Reply:
Hi @borkmann Thank you so much for making it clearer I was in a wrong path since the beginning. I still have some doubt. Replacing individual calls by direct access (ctx->data and ctx->data_end) I would use a direct assignment with a correspondent casting for destination structure instead of using memcpy. Doing this way, the non-linear data parts will be covered anyway?

Reply:
See for example how https://github.com/cilium/cilium/blob/master/bpf/lib/nodeport.h#L201 data/data_end & `revalidate_data()` is used to assign the ip6 pointer and then read from it.

Reply:
Thanks for sharing. I'm working on it.